flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferBuilder.java

```diff
@@ -153,7 +153,12 @@ public BufferRecycler getRecycler() {
     }
 
     public void recycle() {
-        recycler.recycle(memorySegment);
+        // If at least one consumer was created then they responsible for the memory recycling
+        // because BufferBuilder doesn't contain a references counter so it will be impossible to
+        // correctly recycle memory here.
+        if (!bufferConsumerCreated) {
+            recycler.recycle(memorySegment);
+        }
```

Comment:
Frankly, this seems (still?) like a partial solution/hack, because what's the contract when this method should be called? Before this `recycle()` method was introduced, it was at least clear, that `BufferBuilder` is never recycling the segment, and this is always done via closing `BufferConsumer`s. Now it seems like we are going deeper into the murky waters that "sometimes" `recycle()` should be called?

I think I would like the "hack" of relaying on the `bufferConsumerCreated` flag to avoid retaining the buffer in the `BufferBuilder`, but I think it's still confusing:
1. It would be probably better to make `BufferBuilder` implement `Closeable` and rename `recycle()` to `close()` - this would probably cause quite a bit of changes, especially in tests.
2. It still doesn't solve the problem of writing to already released `memorySegment`, like what if the `BufferConsumer` was created and has already been closed, while someone is still writing data to the `BufferBuilder`? There was a bug like that, that I fixed on a different layer, but maybe we should fix this after all here as well? I mean especially measuring on the benchmarking machine if there is any overhead of retaining and recycling `NetworkBuffer` one extra time?

Reply:
Yes, it is still definitely a hack. In my opinion, the right solution is to avoid direct writing into `MemorySegment`  from `BufferBuilder`. It means we just should change the implementation of `BufferBuilder` in such a way that using `Buffer` instead of `MemorySegment`. As I understand now you don't have any objections about such a solution if the benchmark doesn't show any degradation?

In any case, some answers to the questions:

- The contract is simple - `BufferBuilder#recycle()` should be called always when `BufferBuilder` is not needed anymore. You don't need to think `BufferConsumer` was created or not.
- In general, renaming `recycle()` to `close()` makes sense to me since `BufferBuilder` doesn't have `retain` method and ideally, should be closed after usage.(we can think about it when we will agree on a final solution)
- There are a couple of problems still not resolved - writing to already released `memorySegment` or creating 'BufferConsumer' from already closed 'BufferBuilder'. They both can be resolved by solution which we already discussed(using `Buffer` instead of `memorySegment` inside of `BufferBuilder`)

Reply:
I have investigated the code a little more and now I understand that it is not so easy to rewrite `BufferBuilder`(at least I see pretty strange usage of `BufferBuilder` inside of `PartitionSortedBuffer` and `SortMergeResultPartition`). So I can propose for this PR to remove the first commit because it is too controversial and merge the second commit as is - it is not the solution for all problems but it is resolving one of them. And I also suggest creating a ticket where we can think about the correct usage of `MemorySegment` and remove it from the places where it doesn't make sense in order to keep the contract simple and consistent.
Also, if we think that this ticket is not so urgent, we can discuss `MemorySegment` usage here rather than in a separate ticket.
 WDYT? @pnowojski